### ISSUE TITLE
Switch to Hessian-vector products for approximating Fisher vector products in `make_empirical_fisher_vp`

### ISSUE THREAD

**Issue by agrawalraj** [BasisResearch/chirho#428]
Following the discussion in #407, the score outer product formulation for computing Fisher information results in a non-symmetric empirical Fisher matrix. Since conjugate gradients assumes a symmetric matrix, this choice of approximation leads to a lack of composability for compute inverse Fisher vector products. To address this issue, @eb8680 suggested using the Hessian formulation of Fisher. We can do the Hessian formulation as follows using the Pearlmutter trick:

```python
def make_empirical_fisher_vp(
    func_log_prob: Callable[Concatenate[ParamDict, Point[T], P], torch.Tensor],
    log_prob_params: ParamDict,
    data: Point[T],
    *args: P.args,
    **kwargs: P.kwargs,
) -> Callable[[ParamDict], ParamDict]:
    N = data[next(iter(data))].shape[0]
    mean_vector = 1 / N * torch.ones(N)

    def bound_batched_func_log_prob(params: ParamDict) -> torch.Tensor:
        return batched_func_log_prob(params, data)

    def _empirical_fisher_vp(v: ParamDict) -> ParamDict:
        jvp_fn = lambda log_prob_params: torch.func.jvp(
            bound_batched_func_log_prob, (log_prob_params,), (v,)
        )[1]
        vjp_fn = torch.func.vjp(jvp_fn, log_prob_params)[1]
        return vjp_fn(-1 * mean_vector)[0]  # Fisher = -E[Hessian]

    return _empirical_fisher_vp
```

I also think we should pass in an optimal jitter argument to handle the case when the empirical fisher information matrix is non-invertible (e.g., potentially when # monte carlo samples is less than the dimension of the guide parameters). For conjugate gradients, this jitter might not be needed since the the minimum norm solution is returned for singular matrices: https://scicomp.stackexchange.com/questions/35239/what-happens-when-i-use-a-conjugate-gradient-solver-with-a-symmetric-positive-se
